### PR TITLE
[rustdoc] Add `--output-format=doctest` command-line flag

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of https://github.com/rust-lang/rust/issues/134529.

It was discussed with the Rust-for-Linux project recently that they needed a way to extract doctests so they can modify them and then run them more easily (look for "a way to extract doctests" [here](https://github.com/Rust-for-Linux/linux/issues/2)).

For now, I output most of `ScrapedDoctest` fields in JSON format with `serde_json`. So it outputs the following information:

 * filename
 * line
 * langstr
 * text

cc @ojeda 
r? @notriddle

### CODE REVIEW

**Comment by ojeda** [comment]
That was quick, thanks a lot!

@rustbot label A-rust-for-linux

File: src/doc/rustdoc/src/unstable-features.md [hunk 1]
```diff
@@ -664,3 +664,22 @@ Similar to cargo `build.rustc-wrapper` option, this flag takes a `rustc` wrapper
...
+ * File where they are located.
+ * Line where they are located.
+ * Codeblock attributes (more information about this [here](./write-documentation/documentation-tests.html#attributes)).
+
+The output format is JSON.
+
+Using this flag looks like this:
+
+```bash
+$ rustdoc -Zunstable-options --extract-doctests src/lib.rs
```
Comment (notriddle) [inline src/doc/rustdoc/src/unstable-features.md:684]: Why is this a CLI option and not an output format?

File: src/doc/rustdoc/src/unstable-features.md [hunk 2]
```diff
@@ -664,3 +664,22 @@ Similar to cargo `build.rustc-wrapper` option, this flag takes a `rustc` wrapper
...
+ * Tracking issue: [#134529](https://github.com/rust-lang/rust/issues/134529)
+
+When this flag is used, it outputs the doctests original source code alongside information
+such as:
+
+ * File where they are located.
+ * Line where they are located.
+ * Codeblock attributes (more information about this [here](./write-documentation/documentation-tests.html#attributes)).
+
+The output format is JSON.
```
Comment (notriddle) [inline src/doc/rustdoc/src/unstable-features.md:679]: That's not enough documentation, even for an unstable feature. You need to say what the actual keys are, and what they do (particularly the subkeys of `langstr`, which are not self-explanatory). Please include a prettified example JSON document with all of the format features used.

**Review by notriddle — CHANGES_REQUESTED** [review]

**Comment by aDotInTheVoid** [comment]
I don't think the approach taken here (or manually implementing `serde::Serialize` for internal structs) is a good one.

It means some of the internal structs are actually part of the public API, but the exact relation is unclear in source code.

I think the right way to structure this is to do what `--output-format json` does and keep the type definitions for (de)serialization very separate from the all rustc_*/rustdoc internal types, and have them in there own module/crate. This achieves:

- It's very clear when a PR adjusts the schema (whereas at the moment, that's not super apparent)
- We can just use `#[derive(Serialize, Deserialize)]` on the module (rather than having to implement manually)
- We can republish the module onto crates.io, so it's easy for (non-kernel) rust users (eg nextest, etc)
- We can the modules api-docs to describe the schema.

Some other things:
- There should be an option (probably the default) to output to a file.
- There should be a `FORMAT_VERSION` field at the root of the output. This was super critical for rustdoc-json to be able to evolve without breaking users in a really terrible way.

File: tests/rustdoc-ui/extract-doctests.rs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// Test to ensure that it generates expected output for `--extract-doctests` command-line
+// flag.
+
+//@ compile-flags:-Z unstable-options --extract-doctests
+//@ normalize-stdout-test: "tests/rustdoc-ui" -> "$$DIR"
+//@ check-pass
+
```
Comment (aDotInTheVoid) [inline tests/rustdoc-ui/extract-doctests.rs:7]: We probably want a test for how this interacts with:

- Implicit/explicit `fn main()`
- Hidding lines with `/// # use some::path;`

File: src/doc/rustdoc/src/unstable-features.md (hunk 1 above)
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/unstable-features.md:684]: Because I didn't think about it, it's SO MUCH better!

File: src/doc/rustdoc/src/unstable-features.md (hunk 2 above)
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/unstable-features.md:679]: I'm not too sure if we want everything to be documented as docblock attributes list might get longer. Well, I'll list the current one and we can think about it later.

File: tests/rustdoc-ui/extract-doctests.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply tests/rustdoc-ui/extract-doctests.rs:7]: True although currently it's the code as defined in the documentation and no changes on rustdoc side (which I think matches better what rust for linux wants).

Comment (GuillaumeGomez) [reply tests/rustdoc-ui/extract-doctests.rs:7]: Maybe we should provide another field with "rustdoc computed code" where hidden lines and `main` function wrapping are added by rustdoc. Actually that sounds like a very good idea.

**Comment by GuillaumeGomez** [comment]
> I don't think the approach taken here (or manually implementing `serde::Serialize` for internal structs) is a good one.
> 
> It means some of the internal structs are actually part of the public API, but the exact relation is unclear in source code.
> 
> I think the right way to structure this is to do what `--output-format json` does and keep the type definitions for (de)serialization very separate from the all rustc_*/rustdoc internal types, and have them in there own module/crate. This achieves:
> 
>     * It's very clear when a PR adjusts the schema (whereas at the moment, that's not super apparent)
> 
>     * We can just use `#[derive(Serialize, Deserialize)]` on the module (rather than having to implement manually)
> 
>     * We can republish the module onto crates.io, so it's easy for (non-kernel) rust users (eg nextest, etc)
> 
>     * We can the modules api-docs to describe the schema.

I think you're right, I took the easiest way which has the advantage to always keep the output up-to-date since it uses the same types for the code generation. I'll split it out.

> Some other things:
> 
>     * There should be an option (probably the default) to output to a file.

I'm not convinced that adding an option to output to a file is needed. I have these two cases in mind:
 1. From a shell you redirect stdout.
 2. From a program you can just get the stdout once again and then do what you need with it.

Do you think another case I didn't think about maybe?

>     * There should be a `FORMAT_VERSION` field at the root of the output. This was super critical for rustdoc-json to be able to evolve without breaking users in a really terrible way.

Good catch, gonna add it, thanks!

**Comment by aDotInTheVoid** [comment]
> I took the easiest way which has the advantage to always keep the output up-to-date since it uses the same types for the code generation.

I think the nicest way to get this is to unpack the “internal” types exhaustively in the internal->public conversion. This way you’re forced to think about the public repr when internals change, but they’re still seperate and easy to track.

> Do you think another case I didn't think about maybe?

I think the big one is build-systems, which are much more used to chaining commands together via files than stdout.

File: tests/rustdoc-ui/extract-doctests.rs (hunk 3 above)
Comment (ojeda) [reply tests/rustdoc-ui/extract-doctests.rs:7]: > True although currently it's the code as defined in the documentation and no changes on rustdoc side (which I think matches better what rust for linux wants).

Currently we use both hidden lines and the `?` support, i.e. the `# Ok::<..., ...>(...)` syntax (which implies the `fn main()`, from what I understand).

Some of that post-processing may be easy to do by users, like the hidden lines I assume, but if you walk the source or IR or similar to figure out details (like the crate attributes that you move out of `main()`), then it may be harder for end users to replicate that properly without a hack. Perhaps it may be possible to export what `rustdoc` figured about the test, so that users can replicate the post-processing on their side, but customized to their environment.

Having the `rustdoc` computed code sounds fine since it may be enough for some use cases, but e.g. we currently need to convert on the fly the `.unwrap()` in the `?`-using tests into a custom `assert!`.

Comment (ojeda) [reply tests/rustdoc-ui/extract-doctests.rs:7]: So, for instance, if `rustdoc` tells us "this test uses `?`", "these are the crate attributes I would have moved", etc., then users may be able to easily and reliably construct their own wrappers and e.g. do a check instead of an `.unwrap()`.

Of course, for things like crate attributes, it may be best to remove them so that the end user can re-add them where needed, so it wouldn't be the completely "unaltered" source code. But it could be an "adapted" version. Hidden lines should probably be normal lines in that adapted version.

Perhaps it makes sense to provide all those versions, i.e. the completely unaltered one for those that may want to do something complex or to render the text somewhere, the "adapted" version for customized test environments and the `rustdoc` computed code for those that can use directly that. In the kernel we would use the "adapted" one, only, I think.

**Comment by GuillaumeGomez** [comment]
Big update:

I separated output type from rustdoc's type as suggested by @aDotInTheVoid as it will make it much easier for us to maintain.

I added a `format_version` field.

I added the modified rustdoc code as well (in a `doctest_code` field).

I changed the option to `--output-format=doctest`.

I greatly improved the book description.

Hopefully I didn't miss anything. :)

**Comment by GuillaumeGomez** [comment]
And fixed `fmt`.

File: src/librustdoc/config.rs [hunk 4]
```diff
@@ -446,12 +448,22 @@ impl Options {
         }
 
         // check for `--output-format=json`
-        if !matches!(matches.opt_str("output-format").as_deref(), None | Some("html"))
+        if let Some(format) = matches.opt_str("output-format").as_deref()
+            && format != "html"
             && !matches.opt_present("show-coverage")
```
Comment (aDotInTheVoid) [inline src/librustdoc/config.rs:453]: What happens if I run `rustdoc --output-format=doctest --show-coverage`

File: src/doc/rustdoc/src/unstable-features.md [hunk 5]
```diff
@@ -542,6 +544,67 @@ It can also be used with `--show-coverage`. Take a look at its
...
+For this rust code:
+
+```rust
+/// ```
+/// let x = 12;
+/// ```
+pub trait Trait {}
+```
+
+The generated output will look like this:
```
Comment (aDotInTheVoid) [inline src/doc/rustdoc/src/unstable-features.md:568]: This should document that the output goes to `stdout`

File: src/librustdoc/doctest.rs [hunk 6]
```diff
@@ -226,21 +228,42 @@ pub(crate) fn run(dcx: DiagCtxtHandle<'_>, input: Input, options: RustdocOptions
...
+                tests.into_iter().for_each(|t| collector.add_test(t));
+
+                Ok(Some(collector))
+            }
         });
         compiler.sess.dcx().abort_if_errors();
 
         collector
-    });
+    }) {
```
Comment (aDotInTheVoid) [inline src/librustdoc/doctest.rs:259]: This control-flow is hard to follow. Given that the only source of `Ok(None)` and `Err` are in the `exctract_doctests` branch, the early return should be moved up there (so when reading them there it's clear that they end the function)

File: src/librustdoc/doctest/extracted.rs [hunk 7]
```diff
@@ -0,0 +1,132 @@
+use serde::Serialize;
```
Comment (aDotInTheVoid) [inline src/librustdoc/doctest/extracted.rs:1]: There should be a doc-comment here explaining what feature this is supporting, and that `FORMAT_VERSION` needs to be bumped when the types are changed.

File: src/doc/rustdoc/src/unstable-features.md [hunk 8]
```diff
@@ -542,6 +544,67 @@ It can also be used with `--show-coverage`. Take a look at its
...
+}
+```
+
+ * `format_version` gives you the current version of the generated JSON. If we change the output in any way, the number will increase.
+ * `doctests` contains the list of doctests present in the crate.
+   * `file` is the file path where the doctest is located.
+   * `line` is the line where the doctest starts (so where the \`\`\` is located in the current code).
+   * `doctest_attributes` contains computed information about the attributes used on the doctests. For more information about doctest attributes, take a look [here](write-documentation/documentation-tests.html#attributes).
+   * `original_code` is the code as written in the source code before rustdoc modifies it.
+   * `doctest_code` is the code modified by rustdoc that will be run. If there is a syntax error, this field will not be present.
```
Comment (aDotInTheVoid) [inline src/doc/rustdoc/src/unstable-features.md:605]: This isn't right: From the doctest, we output `"doctest_code":"#![allow(unused)]\nfn main() {\nlet\n}"`

File: tests/rustdoc-ui/extract-doctests.stdout [hunk 9]
```diff
@@ -0,0 +1 @@
+{"format_version":1,"doctests":[{"file":"$DIR/extract-doctests.rs","line":8,"doctest_attributes":{"original":"ignore (checking attributes)","should_panic":false,"no_run":false,"ignore":"All","rust":true,"test_harness":false,"compile_fail":false,"standalone_crate":false,"error_codes":[],"edition":null,"added_css_classes":[],"unknown":[]},"original_code":"let x = 12;\nlet y = 14;","doctest_code":"#![allow(unused)]\nfn main() {\nlet x = 12;\nlet y = 14;\n}","name":"$DIR/extract-doctests.rs - (line 8)"},{"file":"$DIR/extract-doctests.rs","line":13,"doctest_attributes":{"original":"edition2018,compile_fail","should_panic":false,"no_run":true,"ignore":"None","rust":true,"test_harness":false,"compile_fail":true,"standalone_crate":false,"error_codes":[],"edition":"2018","added_css_classes":[],"unknown":[]},"original_code":"let","doctest_code":"#![allow(unused)]\nfn main() {\nlet\n}","name":"$DIR/extract-doctests.rs - (line 13)"}]}
```
Comment (aDotInTheVoid) [inline tests/rustdoc-ui/extract-doctests.stdout:1]: I still don't love this as a way of testing, but it's probably fine for now. Eventually, `run-make` is probably nicer.

File: src/librustdoc/lib.rs [hunk 10]
```diff
@@ -814,7 +814,12 @@ fn main_args(early_dcx: &mut EarlyDiagCtxt, at_args: &[String]) {
...
 
-    match (options.should_test, config::markdown_input(&input)) {
+    let output_format = options.output_format;
+
+    match (
+        options.should_test || output_format == config::OutputFormat::Doctest,
+        config::markdown_input(&input),
+    ) {
         (true, Some(_)) => return wrap_return(dcx, doctest::test_markdown(&input, options)),
         (true, None) => return doctest::run(dcx, input, options),
```
Comment (aDotInTheVoid) [inline src/librustdoc/lib.rs:824]: `doctest::run` should be renamed, because now it either runs *or* just collects and prints the doctests.

**Review by aDotInTheVoid — COMMENTED** [review]
This is looking in alot better shape, thanks.

I think it's fine to land on nightly while the public API uses the internal types, I'd like to move away from that eventually, but that can be done in a follow-up PR. Same with test improvements.


@notriddle Are you happy for me to r+ this after I'm happy with with it, or should I also make sure you're good with the current version before merging?

File: src/librustdoc/config.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustdoc/config.rs:453]: Good point! Gonna also add a UI test for this case.

File: src/librustdoc/doctest.rs (hunk 6 above)
Comment (GuillaumeGomez) [reply src/librustdoc/doctest.rs:259]: We can't because it'd leave the closure, not the whole `run` function. Hence why I had to do it in two steps. I can improve the code by storing the `run_compiler` returned value in a variable though.

File: src/librustdoc/doctest/extracted.rs (hunk 7 above)
Comment (GuillaumeGomez) [reply src/librustdoc/doctest/extracted.rs:1]: Very good point!

File: src/doc/rustdoc/src/unstable-features.md (hunk 8 above)
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/unstable-features.md:605]: I don't really know how to word it. In this case the AST isn't invalid, it's incomplete (which is not the same for the compiler).

File: tests/rustdoc-ui/extract-doctests.stdout (hunk 9 above)
Comment (GuillaumeGomez) [reply tests/rustdoc-ui/extract-doctests.stdout:1]: I wanted to re-use the `rustdoc-json` testsuite originally but it'd require some extra changes. So for now I have put it here and I plan to send a follow-up where I add an option allowing `rustdoc-json` tests to use `--output-format=doctest` instead.

File: src/librustdoc/lib.rs (hunk 10 above)
Comment (GuillaumeGomez) [reply src/librustdoc/lib.rs:824]: We run the `doctest` pass. Same for outputs named `run_renderer`. I don't mind changing it though, how would you rename it?

**Comment by GuillaumeGomez** [comment]
Applied suggestions.

**Comment by notriddle** [comment]
> Are you happy for me to r+ this after I'm happy with with it, or should I also make sure you're good with the current version before merging?

r=me

**Comment by GuillaumeGomez** [comment]
Then let's go. Thanks to both of you!

@bors r=notriddle,aDotInTheVoid rollup
